README.md
# Bubs

Bubs is a Keras/TensorFlow reimplementation of the Flair Contextualized Embeddings (https://alanakbik.github.io/papers/coling2018.pdf). It was developed as a building block for use in Keras/TensorFlow natural language models by [Yuliya Dovzhenko](https://github.com/ydovzhenko) and the [Kensho Technologies AI Research Team](https://www.kensho.com/) ([full contributor list](https://github.com/kensho-technologies/bubs/blob/master/AUTHORS.md)).

Please check out our [blog post](https://blog.kensho.com/bubs-14c71be43fe9) announcing Bubs!
 
----------------------------------------------

Bubs implements two types of Flair embeddings:  `news-forward-fast` and `news-backward-fast`.

Bubs consists of two parts:
* ContextualizedEmbedding: a Keras custom layer, which computes the contextualized embeddings. It has two outputs corresponding to the `news-forward-fast` and `news-backward-fast` embeddings.
* InputEncoder: an object for constructing inputs to the ContextualizedEmbedding layer.


#### ContextualizedEmbedding
This layer consists of:
* two character-level embedding layers (one to be used as input to the forward LSTM, one to be used as an input to the backward LSTM)
* two character-level LSTM layers (one going forward, one going backward along the sentence) 
* two indexing layers for selecting character-level LSTM outputs at the locations where tokens end/begin (resulting in two output vectors per token).
* two masking layers to make sure the outputs at padded locations are set to zeros. This is necessary because sentences will have different numbers of tokens and the outputs will be padded to max_token_sequence_length.

The following inputs to the ContextualizedEmbedding layer are required:
* `forward_input`: padded array of character codes corresponding to each sentence with special begin/end characters
* `backward_input`: padded array of character codes in reverse order with special begin/end characters
* `forward_index_input`: padded array of locations of token outputs in forward_input
* `backward_index_input`: padded array of locations of token outputs in backward_input
* `forward_mask_input`: mask of same shape as forward_index_input, with 0's where padded and 1's where real tokens
* `backward_mask_input`:mask of same shape as back_index_input, with 0's where padded and 1's where real tokens

#### InputEncoder

This class provides two methods for preparing inputs to the ContextualizedEmbedding layer:

* `input_batches_from_raw_text()` will accept a raw text string, split it into sentences and tokens, enforce character and token limits by breaking longer sentences into parts. It will then translate characters into numeric codes from the dictionary in `char_to_int.py`, pad sentences to the same length, and compute indices of token-level outputs from the character-level LSTMs.

* `prepare_inputs_from_pretokenized()` will accept a list of lists of tokens and output model inputs . Use at your own risk: this function will not enforce character or token limits and will assume that all sentences fit into one batch. Make sure you split all your sentences into batches before calling this function. Otherwise the indices in `forward_index_input` and `backward_index_input` will be incorrect.

### The model weights

The weights of the ContextualizedEmbedding layer were copied from the corresponding weights inside flair's `news-forward-fast` and `news-backward-fast` embeddings (see `scripts/extracting_model_weights.py` for a code snippet that was used to extract the weights).

### The name Bubs

Bubs is named after the author's cat, Bubs (short for Bubbles).

### A minimal example model
Below we define a very simple example that outputs contextualized embeddings for the following text: "Bubs is a cat. Bubs is cute.".

```python
from tensorflow.keras.layers import Input
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam

from bubs import ContextualizedEmbedding, load_weights_from_npz
from bubs.helpers import InputEncoder


MAX_TOKEN_SEQUENCE_LEN = 125
MAX_CHAR_SEQUENCE_LEN = 2500

"""Load the default weights (provided with this package). If you would like to provide your own 
weights, you may pass a path to the weights npz file to the load_weights_from_npz() function.
"""
weights = load_weights_from_npz()
context_embedding_layer = ContextualizedEmbedding(MAX_TOKEN_SEQUENCE_LEN, weights)

"""Required: define inputs to the ContextualizedEmbedding layer"""
forward_input = Input(shape=(None,), name="forward_input", dtype="int16")
backward_input = Input(shape=(None,), name="backward_input", dtype="int16")
forward_index_input = Input(
    batch_shape=(None, MAX_TOKEN_SEQUENCE_LEN, 2), name="forward_index_input", dtype="int32"
)
forward_mask_input = Input(
    batch_shape=(None, MAX_TOKEN_SEQUENCE_LEN), name="forward_mask_input", dtype="float32"
)
backward_index_input = Input(
    batch_shape=(None, MAX_TOKEN_SEQUENCE_LEN, 2), name="backward_index_input", dtype="int32"
)
backward_mask_input = Input(
    batch_shape=(None, MAX_TOKEN_SEQUENCE_LEN), name="backward_mask_input", dtype="float32"
)

all_inputs = [
    forward_input,
    backward_input,
    forward_index_input,
    backward_index_input,
    forward_mask_input,
    backward_mask_input,
]

forward_embeddings, backward_embeddings = context_embedding_layer(all_inputs)

model = Model(inputs=all_inputs, outputs=[forward_embeddings, backward_embeddings])
model.compile(optimizer=Adam(), loss="categorical_crossentropy")
```
Now, let's get contextualized embeddings for each token in a couple of sentences.
```python
# Initialize an InputEncoder for creating model inputs from raw text sentences
input_encoder = InputEncoder(MAX_TOKEN_SEQUENCE_LEN, MAX_CHAR_SEQUENCE_LEN)

# Embed a couple of test sentences
raw_text = "Bubs is a cat. Bubs is cute."

(
    generator,
    num_batches,
    document_index_batches
) = input_encoder.input_batches_from_raw_text(raw_text, batch_size=128)

# Only one batch, so we use the generator once
forward_embedding, backward_embedding = model.predict_on_batch(next(generator))
```

The shape of each output will be (2, 125, 1024) for:
* 2 sentences
* 125 words in a padded sentence = `MAX_TOKEN_SEQUENCE_LEN`
* 1024: dimension of the embedding for each word
    
Note that the outputs are padded with zeros from the left. For example, to get the forward and backward embedding of the word 'Bubs' in the first sentence, you would need to index the following locations in the model outputs:
`forward_embedding[0, -5]` 
`backward_embedding[0, -5]`

The embeddings for the word 'cat' are: `forward_embedding[0, -2]` and 
`backward_embedding[0, -2]`

### Installation

```pip install bubs```

# License

Licensed under the Apache 2.0 License. Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

Copyright 2019 Kensho Technologies, Inc.


CHANGELOG.md
#Changelog
###v0.0.1
Initial release.
###v0.0.2
Add `from_config()` function to the ContextualizedEmbedding object, which is called on deserialization. It turns pickled lists of lists back into original numpy arrays.
###v0.1.0
Change to `tf.keras` instead of keras because it's the future of keras and makes tensorflow version irrelevant.



bubs/tests/test_imports.py
# Copyright 2019 Kensho Technologies, LLC.
"""Fake tests that attempt to at least import a few files."""
import unittest


class TestImports(unittest.TestCase):
    def test_imports(self):
        from ..embedding_layer import ContextualizedEmbedding  # noqa
        from ..helpers import InputEncoder  # noqa


bubs/tests/test_tokenizer.py
# Copyright 2019 Kensho Technologies, LLC.
import unittest

from ..tokenizer import RegexTokenizer


class TestRegexTokenizer(unittest.TestCase):
    def test_regex_tokenizer(self):
        tokenizer = RegexTokenizer()
        # Simple tokenization
        text = "Muffins are delicious."
        expected_tokens = [["Muffins", "are", "delicious", "."]]
        expected_spans = [[(0, 7), (8, 11), (12, 21), (21, 22)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)

        # test long sentence
        text = "catch my tail " * 100
        expected_tokens = [["catch", "my", "tail"] * 100]
        tokens, _ = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)

        # multiple sentences
        text = "Muffins are delicious. I lost my hat."
        expected_tokens = [["Muffins", "are", "delicious", "."], ["I", "lost", "my", "hat", "."]]
        expected_spans = [
            [(0, 7), (8, 11), (12, 21), (21, 22)],
            [(0, 1), (2, 6), (7, 9), (10, 13), (13, 14)],
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # linebreaks also split
        text = "Muffins are\ndelicious."
        expected_tokens = [["Muffins", "are"], ["delicious", "."]]
        expected_spans = [[(0, 7), (8, 11)], [(0, 9), (9, 10)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # unicode linebreaks as well
        text = "Muffins are\x0adelicious."
        expected_tokens = [["Muffins", "are"], ["delicious", "."]]
        expected_spans = [[(0, 7), (8, 11)], [(0, 9), (9, 10)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # edge cases don't raise
        text = ""
        expected_tokens = []
        expected_spans = []
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # proper whitespace stripping including unicode
        text = "\n\n\t\x0a\t \u3000\t"
        expected_tokens = []
        expected_spans = []
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # combination of many things
        text = " \n ''Sentence\" \tone. Another \u201cWEIRD sentence  !\t\t"
        expected_tokens = [
            ["''", "Sentence", '"', "one", "."],
            ["Another", "\u201c", "WEIRD", "sentence", "!"],
        ]
        expected_spans = [
            [(0, 2), (2, 10), (10, 11), (13, 16), (16, 17)],
            [(0, 7), (8, 9), (9, 14), (15, 23), (25, 26)],
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # sentence terminal tokenization
        text = "This is a ?sentence,. This is another sentence?,"
        expected_tokens = [
            ["This", "is", "a", "?", "sentence", ",", "."],
            ["This", "is", "another", "sentence", "?", ","],
        ]
        expected_spans = [
            [(0, 4), (5, 7), (8, 9), (10, 11), (11, 19), (19, 20), (20, 21)],
            [(0, 4), (5, 7), (8, 15), (16, 24), (24, 25), (25, 26)],
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        #  Test random html escapes
        text = "&amp; sentence"
        expected_tokens = [["&", "amp", ";", "sentence"]]
        expected_spans = [[(0, 1), (1, 4), (4, 5), (6, 14)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # test that uris are properly split
        text = "http://www.foo.com/&amp;bar.html plus a sentence"
        expected_tokens = [["http://www.foo.com/&amp;bar.html", "plus", "a", "sentence"]]
        expected_spans = [[(0, 32), (33, 37), (38, 39), (40, 48)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # test possessive splitting
        text = "aron's duck"
        expected_tokens = [["aron", "'s", "duck"]]
        expected_spans = [[(0, 4), (4, 6), (7, 11)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        text = "kittens' delight"
        expected_tokens = [["kittens", "'", "delight"]]
        expected_spans = [[(0, 7), (7, 8), (9, 16)]]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # test contraction splitting
        text = "kittens you'd think wouldn't do that!"
        expected_tokens = [["kittens", "you", "'d", "think", "would", "n't", "do", "that", "!"]]
        expected_spans = [
            [(0, 7), (8, 11), (11, 13), (14, 19), (20, 25), (25, 28), (29, 31), (32, 36), (36, 37)]
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # Test tokenize() with max_characters_per_token
        tokenizer = RegexTokenizer(max_characters_per_token=5)
        text = "kittens you'd think wouldn't do that!"
        expected_tokens = [["kitte", "ns", "you", "'d", "think", "would", "n't", "do", "that", "!"]]
        expected_spans = [
            [
                (0, 5),
                (5, 7),
                (8, 11),
                (11, 13),
                (14, 19),
                (20, 25),
                (25, 28),
                (29, 31),
                (32, 36),
                (36, 37),
            ]
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        expected_tokens = [["kitte", "ns", "you", "'d", "think", "would", "n't", "do", "that", "!"]]
        expected_spans = [
            [
                (0, 5),
                (5, 7),
                (8, 11),
                (11, 13),
                (14, 19),
                (20, 25),
                (25, 28),
                (29, 31),
                (32, 36),
                (36, 37),
            ]
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # extreme case of max_characters_per_token = 1
        tokenizer = RegexTokenizer(max_characters_per_token=1)
        text = "kittens you'd think wouldn't do that!"
        expected_tokens = [
            [
                "k",
                "i",
                "t",
                "t",
                "e",
                "n",
                "s",
                "y",
                "o",
                "u",
                "'",
                "d",
                "t",
                "h",
                "i",
                "n",
                "k",
                "w",
                "o",
                "u",
                "l",
                "d",
                "n",
                "'",
                "t",
                "d",
                "o",
                "t",
                "h",
                "a",
                "t",
                "!",
            ]
        ]
        expected_spans = [
            [
                (0, 1),
                (1, 2),
                (2, 3),
                (3, 4),
                (4, 5),
                (5, 6),
                (6, 7),
                (8, 9),
                (9, 10),
                (10, 11),
                (11, 12),
                (12, 13),
                (14, 15),
                (15, 16),
                (16, 17),
                (17, 18),
                (18, 19),
                (20, 21),
                (21, 22),
                (22, 23),
                (23, 24),
                (24, 25),
                (25, 26),
                (26, 27),
                (27, 28),
                (29, 30),
                (30, 31),
                (32, 33),
                (33, 34),
                (34, 35),
                (35, 36),
                (36, 37),
            ]
        ]
        tokens, spans = tokenizer.tokenize(text)
        self.assertEqual(expected_tokens, tokens)
        self.assertEqual(expected_spans, spans)

        # Test that an error is thrown if max_characters_per_token is not an integer
        with self.assertRaises(ValueError):
            tokenizer = RegexTokenizer(max_characters_per_token=0.1)

        # Test than an error is thrown if max_characters_per_token is less than 1
        with self.assertRaises(ValueError):
            tokenizer = RegexTokenizer(max_characters_per_token=0)

        # Test than an error is thrown if max_characters_per_token is negative
        with self.assertRaises(ValueError):
            tokenizer = RegexTokenizer(max_characters_per_token=-1)


bubs/tokenizer.py
# Copyright 2019 Kensho Technologies, LLC.
import funcy
from segtok.segmenter import split_single, to_unix_linebreaks
from segtok.tokenizer import (
    split_contractions, split_possessive_markers, web_tokenizer, word_tokenizer
)


def _html_tokenize(sentence):
    """Tokenize string into words, not splitting URIS or emails, wrapping segtok:word_tokenizer.

    It does not split URIs or e-mail addresses. It does not treat html escapes
    as single characters outside of these instances.(eg. &amp; -> '&', 'amp', ';')

    Args:
        sentence: input string for tokenization

    Returns:
        tokens: list of str
    """
    tokens = []
    for i, span in enumerate(web_tokenizer.split(sentence)):
        if i % 2:
            tokens.append(span)
        else:
            tokens.extend(word_tokenizer(span))
    return tokens


class RegexTokenizer:
    """Fast regex-based tokenizer, wrapping around https://github.com/fnl/segtok ."""

    def __init__(self, max_characters_per_token=None):
        """Set up tokenizer specific flags.

        Args:
            max_characters_per_token: maximum allowed token length, token will be split if too long
        """
        if max_characters_per_token is not None and max_characters_per_token < 1:
            raise ValueError(
                "Maximum number of characters per token cannot be less than 1.\n"
                "Max characters requested:{}".format(max_characters_per_token)
            )
        self._max_characters_per_token = max_characters_per_token

    def word_tokenize(self, text):
        """Get list of string tokens from input string.

        Args:
            text: input string for tokenization
        Yields:
            token: str, non-whitespace tokens
        """
        for token in split_possessive_markers(split_contractions(_html_tokenize(text))):
            if self._max_characters_per_token is not None:
                for token_chunk in funcy.chunks(self._max_characters_per_token, token):
                    yield token_chunk
            else:
                yield token

    def sentence_tokenize(self, text):
        """Get list of string sentences from input string.

        Args:
            text: raw input string
        Yields:
            str: non-whitespace, non-empty sentence strings
        """
        for sentence in split_single(to_unix_linebreaks(text)):
            clean_sentence = sentence.strip()
            if len(clean_sentence) > 0:
                yield clean_sentence

    def tokenize(self, text):
        """Tokenize string into sentences and words and return list of lists of str.

        Args:
            text: raw input string
        Returns:
            tokenized_text: list of str tokens
            token_spans: list of (start, end) tuples specifying token locations in text
        """
        tokenized_text = []
        token_spans = []
        for sentence in self.sentence_tokenize(text):
            tokenized_sentence = list(self.word_tokenize(sentence))
            # further split if we have a max sentence length
            tokenized_text.append(tokenized_sentence)
            token_spans.append(_align_multi_token_lists(sentence, [tokenized_sentence]))
        return tokenized_text, token_spans


def _align_multi_token_lists(text, tokenized_document):
    """Align lists of tokenized sentences to text and return lists of token spans."""
    char_offset = 0
    token_spans = []
    first_sentence = True
    for tokenized_sentence in tokenized_document:
        # use first sentence indicator for adding sentence boundary tokens
        if first_sentence:
            first_sentence = False
        char_offset, parsed_sentence_tokens = _align_token_list(
            text, tokenized_sentence, char_offset
        )
        token_spans.extend(parsed_sentence_tokens)
    # add trailing whitespace token if necessary
    if char_offset < len(text):
        token_spans.append((char_offset, char_offset + len(text)))
    return token_spans


def _align_token_list(text, token_list, char_offset=0):
    """Align list of string tokens to text and return list of Token objects."""
    token_spans = []
    for text_token in token_list:
        start = text.index(text_token, char_offset)
        token_spans.append((start, start + len(text_token)))
        char_offset = start + len(text_token)
    return char_offset, token_spans


scripts/extracting_model_weights.py
# Copyright 2019 Kensho Technologies, LLC.
"""For your reference only: not required for using bubs.

This is a script that shows how model weights were extracted from flair embeddings.
It requires flair: pip install flair before running this.
"""

import numpy as np

from flair.embeddings import FlairEmbeddings


def extract_weights():
    """Copy the exact weights in flair's news-forward-fast and news-backward-fast embeddings."""
    flair_forward_embeddings = FlairEmbeddings("news-forward-fast")
    flair_backward_embeddings = FlairEmbeddings("news-backward-fast")

    """These are the matrices we provide in the flair_news_fast_weights.npz file"""
    char_embed_weights_forward = (
        flair_forward_embeddings.lm.encoder._parameters["weight"].detach().cpu().numpy()
    )
    char_embed_weights_back = (
        flair_backward_embeddings.lm.encoder._parameters["weight"].detach().cpu().numpy()
    )

    weight_input_to_hidden_forward = (
        flair_forward_embeddings.lm.rnn._parameters["weight_ih_l0"].detach().cpu().numpy()
    )
    weight_hidden_to_hidden_forward = (
        flair_forward_embeddings.lm.rnn._parameters["weight_hh_l0"].detach().cpu().numpy()
    )
    bias_input_to_hidden_forward = (
        flair_forward_embeddings.lm.rnn._parameters["bias_ih_l0"].detach().cpu().numpy()
    )
    bias_hidden_to_hidden_forward = (
        flair_forward_embeddings.lm.rnn._parameters["bias_hh_l0"].detach().cpu().numpy()
    )
    weight_input_to_hidden_back = (
        flair_backward_embeddings.lm.rnn._parameters["weight_ih_l0"].detach().cpu().numpy()
    )
    weight_hidden_to_hidden_back = (
        flair_backward_embeddings.lm.rnn._parameters["weight_hh_l0"].detach().cpu().numpy()
    )
    bias_input_to_hidden_back = (
        flair_backward_embeddings.lm.rnn._parameters["bias_ih_l0"].detach().cpu().numpy()
    )
    bias_hidden_to_hidden_back = (
        flair_backward_embeddings.lm.rnn._parameters["bias_hh_l0"].detach().cpu().numpy()
    )

    filename = "flair_news_fast_weights.npz"
    np.savez_compressed(
        filename,
        weight_input_to_hidden_forward=weight_input_to_hidden_forward,
        weight_hidden_to_hidden_forward=weight_hidden_to_hidden_forward,
        bias_input_to_hidden_forward=bias_input_to_hidden_forward,
        bias_hidden_to_hidden_forward=bias_hidden_to_hidden_forward,
        weight_input_to_hidden_back=weight_input_to_hidden_back,
        weight_hidden_to_hidden_back=weight_hidden_to_hidden_back,
        bias_input_to_hidden_back=bias_input_to_hidden_back,
        bias_hidden_to_hidden_back=bias_hidden_to_hidden_back,
        char_embed_weights_forward=char_embed_weights_forward,
        char_embed_weights_back=char_embed_weights_back,
    )


if __name__ == "__main__":
    extract_weights()


bubs/char_to_int.py
# Copyright 2019 Kensho Technologies, LLC.
"""Mapping characters to ints to use as input to the character-level embedding layer."""

"""Ordering of characters in this dict follows that in https://github.com/zalandoresearch/flair
Any character not included will get mapped to <unk>, which in turn gets mapped to code 0."""
CHAR_TO_INT = {  # o
    "<unk>": 0,
    " ": 1,
    "e": 2,
    "n": 3,
    "i": 4,
    "t": 5,
    "a": 6,
    "r": 7,
    "s": 8,
    "o": 9,
    "h": 10,
    "d": 11,
    "l": 12,
    "u": 13,
    "c": 14,
    "m": 15,
    "g": 16,
    "\n": 17,
    ".": 18,
    "f": 19,
    "w": 20,
    "b": 21,
    "p": 22,
    ",": 23,
    "y": 24,
    "k": 25,
    "0": 26,
    "v": 27,
    "1": 28,
    "2": 29,
    "S": 30,
    "-": 31,
    "z": 32,
    ":": 33,
    "A": 34,
    "/": 35,
    "I": 36,
    "T": 37,
    "M": 38,
    "D": 39,
    "B": 40,
    "E": 41,
    "P": 42,
    "3": 43,
    "W": 44,
    "\u00fc": 45,
    "5": 46,
    "4": 47,
    "\u00e4": 48,
    "F": 49,
    ")": 50,
    "C": 51,
    "H": 52,
    "G": 53,
    "(": 54,
    "9": 55,
    "6": 56,
    "R": 57,
    "L": 58,
    "7": 59,
    "N": 60,
    "8": 61,
    "K": 62,
    "'": 63,
    "?": 64,
    '"': 65,
    "x": 66,
    "V": 67,
    "U": 68,
    "O": 69,
    "\u00f6": 70,
    "j": 71,
    "J": 72,
    "!": 73,
    "=": 74,
    "Z": 75,
    "\u00df": 76,
    "q": 77,
    "Y": 78,
    "*": 79,
    "_": 80,
    "+": 81,
    "\u20ac": 82,
    "%": 83,
    ";": 84,
    "]": 85,
    "&": 86,
    "[": 87,
    ">": 88,
    "#": 89,
    "\u2013": 90,
    "\u201c": 91,
    "\u00bb": 92,
    "\u2019": 93,
    "Q": 94,
    "X": 95,
    "\u201e": 96,
    "\u00ab": 97,
    "\u201d": 98,
    "\u00b0": 99,
    "\u00dc": 100,
    "\u00e9": 101,
    "\u00e1": 102,
    "<": 103,
    "\u00b2": 104,
    "\u00c3": 105,
    "\u00ff": 106,
    "$": 107,
    "\u00c7": 108,
    "\u00c4": 109,
    "\u2026": 110,
    "\u00d6": 111,
    "\\": 112,
    "\u2022": 113,
    "\u00ed": 114,
    "{": 115,
    "}": 116,
    "@": 117,
    "\u00a9": 118,
    "\u00e8": 119,
    "\u00bc": 120,
    "\u2018": 121,
    "\u00f3": 122,
    "\u2014": 123,
    "\u00c2": 124,
    "\u00c9": 125,
    "\u00ae": 126,
    "\u00b7": 127,
    "\u00e0": 128,
    "\u00f8": 129,
    "\u00d3": 130,
    "\u00b4": 131,
    "\u00da": 132,
    "\u00e2": 133,
    "~": 134,
    "\u00f1": 135,
    "\u00a4": 136,
    "\u00f0": 137,
    "\u00e5": 138,
    "\u00e7": 139,
    "\u266a": 140,
    "\u00e3": 141,
    "`": 142,
    "\u00b6": 143,
    "\u00a0": 144,
    "\u0161": 145,
    "\u00f4": 146,
    "\u00ee": 147,
    "\u0430": 148,
    "\u00f2": 149,
    "\u043e": 150,
    "\u0438": 151,
    "\u00a7": 152,
    "\u00eb": 153,
    "\u010d": 154,
    "\u00d7": 155,
    "\u2122": 156,
    "\u0101": 157,
    "\u00ec": 158,
    "\u00fa": 159,
    "\u0435": 160,
    "\u00b3": 161,
    "\u0440": 162,
    "\u00ea": 163,
    "\u043d": 164,
    "\u0442": 165,
    "\u203a": 166,
    "\u00ac": 167,
    "\u00e6": 168,
    "\u00fd": 169,
    "\u03bf": 170,
    "\u00a3": 171,
    "\u00d1": 172,
    "\u2020": 173,
    "\u0142": 174,
    "\u00c1": 175,
    "\u0131": 176,
    "\u00ef": 177,
    "\u00c8": 178,
    "\u0107": 179,
    "\u2192": 180,
    "\u017e": 181,
    "\u2032": 182,
    "\u0441": 183,
    "^": 184,
    "\u00c5": 185,
    "\u201a": 186,
    "\u00d8": 187,
    "\u00ce": 188,
    "\u0160": 189,
    "\u00cd": 190,
    "\u0159": 191,
    "\u043a": 192,
    "\u00cf": 193,
    "\u0432": 194,
    "\u043b": 195,
    "\u014d": 196,
    "\u011b": 197,
    "\u00ca": 198,
    "\u015f": 199,
    "\u00fb": 200,
    "\u00b1": 201,
    "\u0421": 202,
    "\u00a1": 203,
    "\u0434": 204,
    "\u00dd": 205,
    "\u043c": 206,
    "\u03b1": 207,
    "\u3002": 208,
    "\u0144": 209,
    "\u010c": 210,
    "\u00bd": 211,
    "\u00f5": 212,
    "\u00d5": 213,
    "\u00cc": 214,
    "\u012b": 215,
    "\u0443": 216,
    "\u0103": 217,
    "\u00c0": 218,
    "\u00ba": 219,
    "\u00d0": 220,
    "\u0431": 221,
    "\u00b5": 222,
    "\u25ba": 223,
    "\ufffd": 224,
    "\u00fe": 225,
    "\u0439": 226,
    "\u0627": 227,
    "\u011f": 228,
    "\u044c": 229,
    "\u016b": 230,
    "\u00f9": 231,
    "\u0437": 232,
    "\u00a8": 233,
    "\u03b2": 234,
    "\u017d": 235,
    "\u2212": 236,
    "\u0151": 237,
    "\u044f": 238,
    "\u0446": 239,
    "\u043f": 240,
    "\u0130": 241,
    "\u2011": 242,
    "\u03b9": 243,
    "\u03bd": 244,
    "\u00bf": 245,
    "\u0433": 246,
    "\u03c4": 247,
    "\u2039": 248,
    "\u03b5": 249,
    "\u00d4": 250,
    "\u00cb": 251,
    "\u0119": 252,
    "\u044b": 253,
    "\u00de": 254,
    "\u00d2": 255,
    "\u015b": 256,
    "\u00aa": 257,
    "\u0219": 258,
    "\u03c1": 259,
    "\u0105": 260,
    "\u0644": 261,
    "\u03bb": 262,
    "\u00b9": 263,
    "\u00f7": 264,
    "\u0449": 265,
    "\u0414": 266,
    "\u0447": 267,
    "\u03c2": 268,
    "\u7684": 269,
    "\u03bc": 270,
    "\u0445": 271,
    "\u016f": 272,
    "\u03c5": 273,
    "\u041f": 274,
}
